Foundational Authentication Service Core

DannyPhlyme opened the pull request:
## ✨ PR Overview: Foundational Authentication Service Core

This pull request introduces the foundational backend logic for the new plug-and-play authentication service (`@super-skeleton/auth`). It lays the groundwork for user management, secure registration, and JWT-based login capabilities.

While this PR incorporates several interconnected foundational pieces to establish a functional baseline, future contributions will aim for more granular, incremental changes as requested.

---

### 🚀 Features Implemented:

This PR delivers the following core components and functionalities:

1.  **👤 Core User Model (`libs/auth/src/lib/user/user.model.ts`)**
    *   Defined the `IUser` interface and `UserModel` (`SsModel`) for Mongoose, enabling user data persistence via the `@super-skeleton/crud` library.
    *   Includes essential fields for identity, credentials, verification status, and timestamps.

2.  **🛡️ Input Validation (`libs/auth/src/lib/user/user.validators.ts`)**
    *   Established Zod schemas (`userRegistrationZodSchema`, `userLoginZodSchema`) for robust validation of registration and login payloads.
    *   Includes derived TypeScript types (`UserRegistrationInput`, `UserLoginInput`) for type safety.

3.  **⚙️ Authentication Service Logic (`libs/auth/src/lib/services/auth.service.ts`)**
    *   Created the `AuthService` class to encapsulate core business logic.
    *   **User Registration (`registerUser` method):**
        *   Securely hashes passwords using **Argon2**.
        *   Checks for existing users to prevent duplicate email registrations.
        *   Creates new user records in the database.
        *   Emits a `userRegistered` event, allowing host applications to hook into the post-registration process (e.g., for sending verification emails). A placeholder verification token is currently part of the event payload.
    *   **User Login (`loginUser` method):**
        *   Securely verifies passwords using **Argon2**.
        *   Finds users by email (with plans to extend to username/email).
        *   Includes basic account status checks (locked, email verified).
        *   Updates user records with `lastLoginAt` and resets `failedLoginAttempts` on successful login; increments `failedLoginAttempts` on failure.
        *   Generates **JWT Access Tokens** (short-lived) and **Refresh Tokens** (longer-lived) using `jsonwebtoken`.
        *   Emits a `userLoggedIn` event.
    *   **Configuration:**
        *   The `AuthService` constructor is configured to accept JWT secrets (`jwtSecret`, `refreshJwtSecret`), promoting secure key management via environment variables in the host application.
    *   **Extensibility Foundation:**
        *   Integrated a Node.js `EventEmitter` for a basic hook/event system, starting with `userRegistered` and `userLoggedIn`.

4.  **🔩 Dependencies:**
    *   Added `argon2` for password hashing.
    *   Added `jsonwebtoken` (and `@types/jsonwebtoken`) for JWT handling.

---

### 🛣️ Next Steps Enabled by this PR:

With this core logic in place, we can now proceed with:
*   Developing the Express controllers and routes to expose the `/signup` and `/login` endpoints.
*   Implementing actual email verification token generation and the verification flow.
*   Building out the password reset functionality.
*   Further developing the hook system for more granular customization.

---

### 📝 Notes for Reviewers:

*   This PR establishes the essential backend services. The corresponding HTTP API endpoints will be built in subsequent PRs.
*   The JWT secrets (`jwtSecret`, `refreshJwtSecret`) are passed to the `AuthService` constructor. The host application will be responsible for providing these, typically from environment variables.
*   Email verification token generation in `registerUser` is currently a placeholder, to be fully implemented next.

Thank you for reviewing!

DannyPhlyme committed:
feat(auth): implement core user model, registration, and login service logic

Implement the foundational components for the authentication service.
This includes:
- Core User model (`IUser`, `UserModel`) for data persistence via `@super-skeleton/crud`.
- Zod schemas (`userRegistrationZodSchema`, `userLoginZodSchema`) for input validation.
- `AuthService` containing core business logic:
  - User registration (`registerUser`):
    - Secure password hashing with Argon2.
    - Duplicate user prevention.
    - User creation and persistence.
    - Emission of `userRegistered` event for extensibility (e.g., email verification).
  - User login (`loginUser`):
    - Secure password verification with Argon2.
    - Generation of JWT-based access and refresh tokens using `jsonwebtoken`.
    - User record updates (last login time, failed login attempts).
    - Emission of `userLoggedIn` event.
- Configuration of `AuthService` to accept JWT secrets.
- Setup of `EventEmitter` within `AuthService` for a basic hook system.
- Addition of necessary dependencies (`argon2`, `jsonwebtoken`).

DannyPhlyme commented:
![AuthService Component](https://github.com/user-attachments/assets/8bc68d8f-ce86-4893-99af-3e9489720ba8)

DannyPhlyme commented:
<img width="1273" alt="Screenshot 2025-05-30 at 2 44 27 AM" src="https://github.com/user-attachments/assets/aeab939d-396b-4e95-ac1e-6ef808ddfc14" />
<img width="1283" alt="Screenshot 2025-05-30 at 2 44 49 AM" src="https://github.com/user-attachments/assets/284e9e86-0f42-470c-bfe8-2d19d5543170" />

consult-with-josh commented:
Let's start by fixing the 7 new issues identified by Sonar

consult-with-josh commented on libs/auth/package.json:
```diff
@@ -3,10 +3,14 @@
     "tslib": "^2.3.0",
-    "@super-skeleton/crud": "0.0.1",
-    "zod": "^3.23.8"
+    "@super-skeleton/crud": "^0.13.0",
```
Change this to "@super-skeleton/crud": "*" as the versions might grow fast

consult-with-josh commented on libs/auth/src/lib/services/auth.service.ts:
```diff
@@ -0,0 +1,198 @@
+		this.refreshTokenExpiry = options.refreshTokenExpiry || '7d';
+	}
+
+	public async registerUser( data: UserRegistrationInput ): Promise<IUser> {
```
Move the functionality for each function into its own separate file.

register-user.function.ts
login-user.function.ts

And so on and so forth.

Inside each of these individual files, break the function down into smaller, logical functions that are properly named and make non-technical sense. Then, call those functions sequentially in this file.

The service file, when read, should always sound like plain English and make sense to a non-technical person. At the end, I expect your register user function to look like this

```typescript
import { preventDuplicateUserRegistration, ... } from './functions'

public async registerUser( data: UserRegistrationInput ): ... {
   await preventDuplicateUserRegistration();
   const password = await hashPassword();
   const user = await createUserRecordInDb();
   const authTokens = await generateAuthenticationTokensForUser();
   this.emit.....
}
```


All other functions should look similar to this structure.

consult-with-josh commented on libs/auth/src/lib/services/auth.service.ts:
```diff
@@ -0,0 +1,198 @@
+	/**
+   * Allows subscribing to events emitted by the AuthService.
+   * @param eventName The name of the event to listen for.
+   * @param listener The callback function to execute when the event is emitted.
```
Remove jsdoc. Instead, name your params descriptively. I think the names already look good enough. There's no need for the docs.

Also, remove this event emitter into its own separate class and let all services extend that class. I suspect you already plan to do this.

consult-with-josh reviewed (changes requested):
Looking good!

I love the approach.

consult-with-josh commented on libs/auth/src/lib/user/user.validators.ts:
```diff
@@ -0,0 +1,23 @@
+import { z } from 'zod';
```
I think It would be sensible to add a request part validator (can validate query, body and params) to the utils library. It can use zod as the background. Such that we can consistently reuse this validator for routes going forward.

consult-with-josh commented on libs/auth/src/lib/user/user.model.ts:
```diff
@@ -0,0 +1,102 @@
+};
+
+export const UserModel: SsModel<IUser> = {
+	name: 'User',
```
Move this hard-coded name to a constant. Preferably, an object that contains the names of all collections. This will reduce the propensity for erroneous names when one is creating references from other collections

DannyPhlyme committed:
refactor(auth): improve service structure, add validation util, and standardize naming

Implements several enhancements based on code review feedback to improve the structure, maintainability, and reusability of the authentication service.

Key changes include:

- Service Layer Refactoring (`AuthService` & Functions):
  - Decomposed primary `AuthService` methods (`registerUser`, `loginUser`) into individual, orchestrated function files within `services/functions/`.
  - Further broke down logic within these new function files into smaller, well-named helper functions for better readability and "plain English" flow.
  - `AuthService` methods now primarily orchestrate calls to these dedicated functions, passing necessary dependencies.

- Base Event Emitter:
  - Introduced `BaseEventEmitterService` to encapsulate common event emitting functionality.
  - `AuthService` now extends `BaseEventEmitterService` for a cleaner event handling mechanism.

- Code Style:
  - Removed JSDoc comments from service methods and functions in favor of descriptive parameter names.

- Reusable Request Validation Utility:
  - Created new generic `validateRequestParts` utility middleware (in `utils/`).
  - This utility uses Zod to validate Express request `body`, `query`, and `params`, providing a consistent validation approach for future route controllers.

- Model Naming Standardization:
  - Introduced `ModelNames` constants (in `constants/`).
  - `UserModel` now uses `ModelNames.USER` for its name, reducing hardcoding and improving consistency for future models and inter-model referencing.

DannyPhlyme committed:
feat(auth): establish clean public API facade and add internal barrel files

Introduces a well-defined public API for the authentication module via `src/index.ts`, abstracting internal implementation details. Also enhances internal module organization by adding missing barrel files to subdirectories.

Key changes include:
- **Public API Facade (`src/index.ts`):**
  - Refactored `src/index.ts` to only re-export public-facing entities.
  - Moved main initialization logic to a new `lib/initialize.ts` (exporting `initializeAuthModule`).
  - Consolidated public types (e.g., `AuthModuleOptions`, `AuthHookUser`) into a new `lib/types.ts`.
  - (Provision made for optional public custom error classes in `lib/errors.ts`).
- **Internal Barrel Files:**
  - Added `index.ts` barrel files to the following internal subdirectories for consistent internal imports:
    - `lib/base/`
    - `lib/constants/`
    - `lib/controllers/`
    - `lib/routes/`
    - `lib/utils`
  - Existing internal barrel files in `services/`, `user/`, `utils/` are maintained.

This structure promotes a minimal and stable public API, improves developer experience for consumers of the library, enhances tree-shaking potential, and aids in internal code maintainability.

DannyPhlyme committed:
feat(auth, api): mount auth module in API and refactor API bootstrap

Integrates the @super-skeleton/auth module into the ss-api application.
Refactors the ss-api bootstrap process for clarity and maintainability by:
- Moving environment variable validation to a dedicated utility.
- Moving AuthModuleOptions configuration to a dedicated utility.
- Mounting the auth router under the '/auth' path in ss-api.
- Adding required JWT secret configurations to ss-api's environment.

DannyPhlyme commented:
I have addressed your changes in commit 9b985b1. Please take a look

DannyPhlyme committed:
fix(crud): Resolve Mongoose schema creation and null query results

Corrects two critical issues in the @super-skeleton/crud library:

1.  **Schema Creation from SsSchema:**
    The `createSchema` function was incorrectly attempting to look up
    Mongoose field definitions directly by field name from model-defaults,
    ignoring the `type` and `extend` properties specified in the SsSchema.
    This led to "Invalid value for schema path" errors in Mongoose when
    a field's type was effectively undefined.

    The function now correctly uses the `SsSchemaFieldTypes` value from
    `SsSchema[field].type` to look up the base Mongoose definition in
    `model-defaults.constants.ts` and merges it with properties from
    `SsSchema[field].extend`.

2.  **Null Query Result Handling in findRecord:**
    The `findRecord` sub-function was not handling cases where
    `model.findOne()` returned `null` (i.e., no document found). It
    attempted to call `.toObject()` on the null result, causing a
    "Cannot read properties of null" TypeError.

    The function now checks for a null result and returns `null` directly,
    allowing upstream functions (like user existence checks) to correctly
    handle cases where a record is not found.

DannyPhlyme committed:
feat(auth): Enhance login and make email verification configurable

Implements several improvements to the login process in @super-skeleton/auth:

1.  Flexible User Identifier for Login:
    The `findUserByEmailOrUsername` function now uses MongoDB's `$or`
    operator, allowing users to log in by providing either their email
    address or their username in the `emailOrUsername` field.

2.  Configurable Email Verification Requirement:
    The `executeLoginUser` function and its supporting infrastructure
    (`AuthService`, `AuthModuleOptions`) now include a
    `requireEmailVerificationForLogin` boolean option.
    - If true (default), login is prevented if the user's email is not verified.
    - If false, login proceeds even if the email is not verified.
    This setting is now managed through `apps/ss-api/src/environment.ts`
    (reading `process.env.AUTH_REQUIRE_EMAIL_VERIFICATION`) and passed
    down via `configureAuthModuleOptions`, centralizing its configuration.

3.  Account Locking on Failed Attempts:
    The `handleFailedLoginAttempt` function now correctly locks a user's
    account (sets `isAccountLocked = true`) if `failedLoginAttempts`
    reaches `MAX_FAILED_LOGIN_ATTEMPTS` (currently 5).

DannyPhlyme committed:
feat(auth): Complete email verification with robust token handling

- Added user email verification:
    - Generates and stores verification token on signup.
    - Provides `/auth/verify-email/:token` endpoint.
    - Verifies token, updates user status, and securely clears token from the database.
- Enhanced underlying @super-skeleton/crud to ensure tokens are properly unset.

DannyPhlyme committed:
feat(api,auth): Implement email sending for user registration verification

- Integrated @super-skeleton/notifications into ss-api to send verification emails upon user signup.
- Added new environment variables to `apps/ss-api/src/environment.ts` for email provider selection (e.g., SendGrid, Mailgun), API keys, sender details, API base URL, and application name.
- Updated `apps/ss-api/src/app/utils/validate-env.util.ts` to validate these new email-related configurations.
- Modified the `onUserSignUp` hook in `apps/ss-api/src/app/utils/configure-auth-module.util.ts`:

DannyPhlyme committed:
feat(auth,api): Implement full password reset flow

- **@super-skeleton/auth:**
  - Added Zod schemas for forgot/reset password inputs.
  - Implemented service logic (`executeRequestPasswordReset`, `executeResetPassword`) for token generation, validation, and password updating.
  - Added `requestPasswordReset` and `resetPassword` methods to `AuthService`.
  - Created `/forgot-password` and `/reset-password` API endpoints with controllers and validation.
  - Defined and registered `onPasswordResetRequested` and `onPasswordResetCompleted` hooks.

- **ss-api:**
  - Added `frontendAppUrl` to environment for reset links.
  - Implemented email sending for `onPasswordResetRequested` (sends a reset link) and `onPasswordResetCompleted` (sends confirmation).

DannyPhlyme committed:
refactor(api): Group environment secrets for better organization

- Restructured the `secrets` object in `apps/ss-api/src/environment.ts` to group settings into `database`, `auth`, `email`, and `app` categories.
- Updated all consuming files to use the new grouped paths for accessing secrets
- This change enhances the clarity and maintainability of environment variable management within the ss-api application.

DannyPhlyme committed:
feat(api): Enhance onUserLogin hook logging

- Updated the `onUserLogin` hook in `auth-module.util.ts` to include the user's ID in the log message for better traceability.

DannyPhlyme committed:
feat(auth): add resend verification email endpoint and logic

- Implemented functionality for users to request a new email verification link.
- Added Zod schema, `AuthService` logic, controller, and route for resending verification emails.
- Included a new `onVerificationEmailResent` hook in `AuthModuleOptions`.
- Integrated email sending for this hook in `ss-api`.

DannyPhlyme committed:
feat(auth): add login with Google endpoint and logic

DannyPhlyme committed:
feat(oauth-client): implement OAuth client with Google authentication

- Create React application with Vite and Tailwind CSS
- Set up proxy configuration for backend communication
- Implement login page with Google OAuth button
- Add protected dashboard with user profile display
- Create auth context for managing authentication state
- Include refresh token in logout request body
- Add ngrok browser warning bypass header
- Configure classic beautiful design system with custom colors and typography

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(oauth-client): update prop type and remove redundant role attributes

- Changed `title` prop to `Readonly` type for improved immutability.
- Removed unnecessary `role="img"` attributes from multiple `<svg>` elements to clean up markup.

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(oauth-client): clean up layout spacing and markup

- Adjusted spacing in welcome message for better readability.
- Removed inconsistent spacing in `Documentation`, `Blog`, `YouTube channel`, `Interactive tutorials`, and `Nx is open source` sections.

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(oauth-client): add consistent spacing around inline JSX elements

- Introduced `{' '}` spacing to fix inconsistent layout and improve readability in the `nx-welcome` component.

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(oauth-client): add newline characters for consistent spacing in code samples

- Updated `<pre>` blocks in `nx-welcome.tsx` to include `{'\\n'}` for better readability and consistent formatting.

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(api): abstract email utility functions for hooks

- Added reusable helper functions (`createSender`, `prepareEmailCredentials`, `sendEmail`) to simplify email sending logic.
- Updated `onUserSignUp`, `onPasswordResetRequested`, `onPasswordResetCompleted`, and `onVerificationEmailResent` hooks to leverage the new utility functions.
- Improved log messages for better debugging clarity.

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(api): modularize environment variable validation logic

- Extracted validation checks into dedicated functions for better readability and maintainability.
- Introduced `exitWithError` for consistent error handling.

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(auth): replace `||` with `??` for more accurate nullish coalescing

- Updated multiple files to use the nullish coalescing operator (`??`) instead of logical OR (`||`) to correctly handle `null` and `undefined` values.
- Improved code clarity and ensured proper fallback behavior where needed.

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(auth): replace null/undefined checks with optional chaining

- Updated multiple files to use optional chaining (`?.`) for cleaner and more concise null/undefined handling.
- Improved code readability and consistency across authentication-related services and middleware.

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(auth): optimize context with memoization and callbacks

- Replaced inline functions with `useCallback` and `useMemo` for improved performance and better dependency management.
- Adjusted `useEffect` dependencies for consistent behavior.

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(auth): remove unused `user` destructuring from Google login response

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(auth): remove redundant non-null assertions and adjust default port

- Eliminated unnecessary `!` non-null assertions across multiple files to enhance code readability.
- Updated OAuth client preview port from `4212` to `4200` for consistency.

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme committed:
refactor(auth): simplify nullish checks and regex patterns

- Replaced verbose nullish checks with concise `??` operator for improved readability and consistency.
- Simplified username validation regex by replacing `/^[a-zA-Z0-9_]+$/` with `/^\w+$/`.
- Updated comment on account lock event for clarity.

Signed-off-by: Daniel Olajumoke <42919578+DannyPhlyme@users.noreply.github.com>

DannyPhlyme commented:
@joshuaben-seth 
I've addressed all SonarQube issues
